### ISSUE TITLE
Crash in C++ Pennant

### ISSUE THREAD

**Issue by elliottslaughter** [StanfordLegion/legion#1407]
I'm running the C++ Pennant application `master` branch, which I understand is the version designated for use with `control_replication`.

Command line:

```
LD_LIBRARY_PATH=$PWD ./pennant -n 4 -f test/sedovsmall/sedovsmall.pnt
```

Crash:

```
pennant: /scratch2/eslaught/legion_repro_for_resilience/pennant/../runtime/legion/legion_trace.cc:2221: void Legion::Internal::TraceViewSet::insert(Legion::Internal::LogicalView*, Legion::Internal::IndexSpaceExpression*, const FieldMask&, bool): Assertion `expr_volume <= total_expr->get_volume()' failed.
```

Backtrace:

```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff355f859 in __GI_abort () at abort.c:79
#2  0x00007ffff355f729 in __assert_fail_base (fmt=0x7ffff36f5588 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=0x7ffff7854268 "expr_volume <= total_expr->get_volume()", 
    file=0x7ffff78521b0 "/scratch2/eslaught/legion_repro_for_resilience/pennant/../runtime/legion/legion_trace.cc", line=2221, function=<optimized out>) at assert.c:92
#3  0x00007ffff3570fd6 in __GI___assert_fail (assertion=0x7ffff7854268 "expr_volume <= total_expr->get_volume()", 
    file=0x7ffff78521b0 "/scratch2/eslaught/legion_repro_for_resilience/pennant/../runtime/legion/legion_trace.cc", line=2221, 
    function=0x7ffff78541d8 "void Legion::Internal::TraceViewSet::insert(Legion::Internal::LogicalView*, Legion::Internal::IndexSpaceExpression*, const FieldMask&, bool)") at assert.c:101
#4  0x00007ffff6cd6f3a in Legion::Internal::TraceViewSet::insert (this=0x7fffbc2485c0, view=0x7fffc405c690, expr=0x7fffc4072610, mask=..., antialiased=false)
    at /scratch2/eslaught/legion_repro_for_resilience/pennant/../runtime/legion/legion_trace.cc:2221
#5  0x00007ffff6ecbaf1 in Legion::Internal::EquivalenceSet::update_tracing_valid_views (this=0x7fffc48bea20, view=0x7fffc405c690, expr=0x7fffc4072610, usage=..., user_mask=..., invalidates=false)
    at /scratch2/eslaught/legion_repro_for_resilience/pennant/../runtime/legion/legion_analysis.cc:17092
#6  0x00007ffff6eab2eb in Legion::Internal::EquivalenceSet::update_set (this=0x7fffc48bea20, analysis=..., expr=0x7fffc4072610, expr_covers=true, user_mask=..., Python Exception <class 'AttributeError'> '
NoneType' object has no attribute 'pointer': 
deferral_events=std::set with 0 elements, 
    Python Exception <class 'AttributeError'> 'NoneType' object has no attribute 'pointer': 
applied_events=std::set with 0 elements, already_deferred=false) at /scratch2/eslaught/legion_repro_for_resilience/pennant/../runtime/legion/legion_analysis.cc:11685
#7  0x00007ffff6e91b90 in Legion::Internal::UpdateAnalysis::perform_analysis (this=0x7fffc48bfd80, set=0x7fffc48bea20, expr=0x7fffc4072610, expr_covers=true, mask=..., Python Exception <class 'AttributeEr
ror'> 'NoneType' object has no attribute 'pointer': 

    deferral_events=std::set with 0 elements, Python Exception <class 'AttributeError'> 'NoneType' object has no attribute 'pointer': 
applied_events=std::set with 0 elements, already_deferred=false)
    at /scratch2/eslaught/legion_repro_for_resilience/pennant/../runtime/legion/legion_analysis.cc:7325
#8  0x00007ffff6ea479f in Legion::Internal::EquivalenceSet::analyze (this=0x7fffc48bea20, analysis=..., expr=0x7fffc4072610, expr_covers=true, traversal_mask=..., Python Exception <class 'AttributeError'>
 'NoneType' object has no attribute 'pointer': 

    deferral_events=std::set with 0 elements, Python Exception <class 'AttributeError'> 'NoneType' object has no attribute 'pointer': 
applied_events=std::set with 0 elements, already_deferred=false)
    at /scratch2/eslaught/legion_repro_for_resilience/pennant/../runtime/legion/legion_analysis.cc:10621
#9  0x00007ffff6e88b17 in Legion::Internal::PhysicalAnalysis::analyze (this=0x7fffc48bfd80, set=0x7fffc48bea20, mask=..., Python Exception <class 'AttributeError'> 'NoneType' object has no attribute 'poin
ter': 
deferral_events=std::set with 0 elements, Python Exception <class 'AttributeError'> 'NoneType' object has no attribute 'pointer': 

    applied_events=std::set with 0 elements, precondition=..., already_deferred=false) at /scratch2/eslaught/legion_repro_for_resilience/pennant/../runtime/legion/legion_analysis.cc:5658
#10 0x00007ffff6e89629 in Legion::Internal::PhysicalAnalysis::perform_traversal (this=0x7fffc48bfd80, precondition=..., info=..., Python Exception <class 'AttributeError'> 'NoneType' object has no attribu
te 'pointer': 
applied_events=std::set with 0 elements)
    at /scratch2/eslaught/legion_repro_for_resilience/pennant/../runtime/legion/legion_analysis.cc:5827
#11 0x00007ffff6e8d74c in Legion::Internal::CollectiveCopyFillAnalysis::perform_traversal (this=0x7fffc48bfd80, precondition=..., version_info=..., Python Exception <class 'AttributeError'> 'NoneType' obj
ect has no attribute 'pointer': 
applied_events=std::set with 0 elements)
    at /scratch2/eslaught/legion_repro_for_resilience/pennant/../runtime/legion/legion_analysis.cc:6583
#12 0x00007ffff7040486 in Legion::Internal::RegionTreeForest::physical_perform_updates (this=0x5555556fb7b0, req=..., version_info=..., op=0x7fffc405fba0, index=0, precondition=..., term_event=..., 
    targets=..., sources=std::vector of length 0, capacity 0, trace_info=..., Python Exception <class 'AttributeError'> 'NoneType' object has no attribute 'pointer': 
map_applied_events=std::set with 0 elements, analysis=@0x7fffc40932f0: 0x7fffc48bfd80, 
    log_name=0x5555556eb160 "calccharlen", uid=1017, collective_rendezvous=false, record_valid=true, check_initialized=true, defer_copies=true)
    at /scratch2/eslaught/legion_repro_for_resilience/pennant/../runtime/legion/region_tree.cc:2150
#13 0x00007ffff6b631b2 in Legion::Internal::SingleTask::map_all_regions (this=0x7fffc405f9f0, must_epoch_op=0x0, defer_args=0x7fffc48b69e0)
    at /scratch2/eslaught/legion_repro_for_resilience/pennant/../runtime/legion/legion_tasks.cc:4115
#14 0x00007ffff6b70637 in Legion::Internal::PointTask::perform_mapping (this=0x7fffc405f9f0, must_epoch_owner=0x0, args=0x7fffc48b69e0)
    at /scratch2/eslaught/legion_repro_for_resilience/pennant/../runtime/legion/legion_tasks.cc:7127
#15 0x00007ffff717c444 in Legion::Internal::Runtime::legion_runtime_task (args=0x7fffc48b69e0, arglen=52, userdata=0x555555842a00, userlen=8, p=...)
    at /scratch2/eslaught/legion_repro_for_resilience/pennant/../runtime/legion/runtime.cc:31923
#16 0x00007ffff43f49cc in Realm::LocalTaskProcessor::execute_task (this=0x5555556f9970, func_id=4, task_args=...)
    at /scratch2/eslaught/legion_repro_for_resilience/pennant/../runtime/realm/proc_impl.cc:1135
#17 0x00007ffff41c5dff in Realm::Task::execute_on_processor (this=0x7fffc48b6860, p=...) at /scratch2/eslaught/legion_repro_for_resilience/pennant/../runtime/realm/tasks.cc:302
```

Frozen process:

```
Process 1520010 on node g0001.stanford.edu is frozen!
```

I'm on control_replication 29e033ee061483c697c5c2fde8e6478691d4efab.

**Comment by lightsighter** [comment]
Pull and try again.

**Comment by elliottslaughter** [comment]
Vanilla Pennant (without checkpointing) now runs with tracing.

With checkpointing, I am now hitting:

```
pennant: /resilience/legion/runtime/legion/runtime.cc:1002: Legion::Internal::RtEvent Legion::Internal::FutureImpl::request_application_instance(Legion::Memory, Legion::Internal::SingleTask*, Legion::UniqueID, Legion::AddressSpaceID, Legion::Internal::ApUserEvent, size_t): Assertion `target.exists()' failed.
```

Backtrace:

```
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=140206336997376) at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=140206336997376) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=140206336997376, signo=signo@entry=6) at ./nptl/pthread_kill.c:89
#3  0x00007f847c50b476 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#4  0x00007f847c4f17f3 in __GI_abort () at ./stdlib/abort.c:79
#5  0x00007f847c4f171b in __assert_fail_base (fmt=0x7f847c6a6150 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=0x7f84804a6dc5 "target.exists()", 
    file=0x7f84804a5c88 "/resilience/legion/runtime/legion/runtime.cc", line=1002, function=<optimized out>) at ./assert/assert.c:92
#6  0x00007f847c502e96 in __GI___assert_fail (assertion=0x7f84804a6dc5 "target.exists()", file=0x7f84804a5c88 "/resilience/legion/runtime/legion/runtime.cc", line=1002, 
    function=0x7f84804a6cf0 "Legion::Internal::RtEvent Legion::Internal::FutureImpl::request_application_instance(Legion::Memory, Legion::Internal::SingleTask*, Legion::UniqueID, Legion::AddressSpaceID, Legion::Internal::ApUserEv"...) at ./assert/assert.c:101
#7  0x00007f847fd7d736 in Legion::Internal::FutureImpl::request_application_instance (this=0x7f84481d9b30, target=..., task=0x7f8448101cf0, task_uid=1291, source=0, inst_ready=..., 
    known_upper_bound_size=18446744073709551615) at /resilience/legion/runtime/legion/runtime.cc:1002
#8  0x00007f847fb29861 in Legion::Internal::SingleTask::trigger_replay (this=0x7f8448101cf0) at /resilience/legion/runtime/legion/legion_tasks.cc:3474
#9  0x00007f847fb39acb in Legion::Internal::IndividualTask::trigger_replay (this=0x7f8448101cf0) at /resilience/legion/runtime/legion/legion_tasks.cc:6760
#10 0x00007f847ff575f4 in Legion::Internal::Memoizable<Legion::Internal::IndividualTask>::initialize_memoizable (this=0x7f8448101cf0) at /resilience/legion/runtime/legion/legion_ops.inl:71
#11 0x00007f847f91db2b in Legion::Internal::Operation::begin_dependence_analysis (this=0x7f8448101ea0) at /resilience/legion/runtime/legion/legion_ops.cc:2278
#12 0x00007f847f91b920 in Legion::Internal::Operation::execute_dependence_analysis (this=0x7f8448101ea0) at /resilience/legion/runtime/legion/legion_ops.cc:1640
#13 0x00007f847f7ca0f9 in Legion::Internal::InnerContext::process_dependence_stage (this=0x7f8450001840) at /resilience/legion/runtime/legion/legion_context.cc:8030
#14 0x00007f847f7da18e in Legion::Internal::InnerContext::handle_dependence_stage (args=0x7f8450259f40) at /resilience/legion/runtime/legion/legion_context.cc:11720
#15 0x00007f847fdfd735 in Legion::Internal::Runtime::legion_runtime_task (args=0x7f8450259f40, arglen=12, userdata=0x560674ea6800, userlen=8, p=...) at /resilience/legion/runtime/legion/runtime.cc:31701
#16 0x00007f847cf8bb6e in Realm::LocalTaskProcessor::execute_task (this=0x560674d63400, func_id=4, task_args=...) at /resilience/legion/runtime/realm/proc_impl.cc:1135
#17 0x00007f847d006ba1 in Realm::Task::execute_on_processor (this=0x7f8450259dc0, p=...) at /resilience/legion/runtime/realm/tasks.cc:302
#18 0x00007f847d00bf0c in Realm::UserThreadTaskScheduler::execute_task (this=0x560674d63740, task=0x7f8450259dc0) at /resilience/legion/runtime/realm/tasks.cc:1632
#19 0x00007f847d009c03 in Realm::ThreadedTaskScheduler::scheduler_loop (this=0x560674d63740) at /resilience/legion/runtime/realm/tasks.cc:1105
#20 0x00007f847d01280e in Realm::Thread::thread_entry_wrapper<Realm::ThreadedTaskScheduler, &Realm::ThreadedTaskScheduler::scheduler_loop> (obj=0x560674d63740)
    at /resilience/legion/runtime/realm/threads.inl:97
#21 0x00007f847d021c93 in Realm::UserThread::uthread_entry () at /resilience/legion/runtime/realm/threads.cc:1355
#22 0x00007f847c523180 in ?? () at ../sysdeps/unix/sysv/linux/x86_64/__start_context.S:90 from /lib/x86_64-linux-gnu/libc.so.6
#23 0x0000000000000000 in ?? ()
```

To be honest, I'm not sure why this would hit a different case with checkpointing. On the initial run-through, the checkpointing wrapper mostly just passes everything through to the runtime as-is, and it shouldn't be launching any operations until you get to the checkpoint itself.

**Comment by lightsighter** [comment]
Do you have a reproducer?

**Comment by elliottslaughter** [comment]
Run:

```console
$ srun -n 1 -N 1 -c 40 -p gpu --exclusive --pty bash --login
$ cd /scratch2/eslaught/resilience/build
$ ./examples/pennant/pennant -n 4 -f ../examples/pennant/test/sedovsmall/sedovsmall.pnt
```

If you need to rebuild, go to `/scratch2/eslaught/resilience/legion/build`, do whatever you need to do in the Legion repo, and build with `make install -j20`.

**Comment by lightsighter** [comment]
This looks like a trace capture violation to me. When the trace is captured the `calc global dt` task only has one future, however on replay it often has multiple futures:

```
(gdb) p this->task_id
$12 = 6800
(gdb) p this->futures
$13 = std::vector of length 1, capacity 1 = {{<Legion::Unserializable<Legion::Future>> = {<No data fields>}, impl = 0x7fffbc078d80}}
(gdb) c
Continuing.
pennant: /scratch2/eslaught/resilience/legion/runtime/legion/runtime.cc:1002: Legion::Internal::RtEvent Legion::Internal::FutureImpl::request_application_instance(Legion::Memory, Legion::Internal::SingleTask*, Legion::UniqueID, Legion::AddressSpaceID, Legion::Internal::ApUserEvent, size_t): Assertion `target.exists()' failed.

Thread 4 "pennant" received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) up 5
#5  0x00007ffff6f7e0a1 in Legion::Internal::SingleTask::trigger_replay (this=0x7fffbc10b6e0)
    at /scratch2/eslaught/resilience/legion/runtime/legion/legion_tasks.cc:3476
3476	                ApUserEvent::NO_AP_USER_EVENT, bounds);
(gdb) p this->task_id
$14 = 6800
(gdb) p this->futures
$15 = std::vector of length 3, capacity 3 = {{<Legion::Unserializable<Legion::Future>> = {<No data fields>}, impl = 0x7fffbc0f0340}, 
  {<Legion::Unserializable<Legion::Future>> = {<No data fields>}, impl = 0x7fffbc0d96d0}, 
  {<Legion::Unserializable<Legion::Future>> = {<No data fields>}, impl = 0x7fffbc10b200}}
  ```

**Comment by elliottslaughter** [comment]
I'm trying to work through this, and it doesn't make any sense to me.

I went into gdb and set breakpoints on `Legion::Runtime::begin_trace(...)`, `Legion::Runtime::end_trace(...)`, and `Legion::Future::Future(Internal::FutureImpl *)`. The latter is, if I understand correctly, the entrypoint where the runtime constructs all user-visible `Future` objects.

When I run through, I see 10 calls to `Future::Future(Internal::FutureImpl *)` on both the first and second pass through the trace. Then I went through again and checked the backtraces on each call, and the look identical.

I'm attaching the full gdb session logs and a summary of just the application-relevant portions of the backtraces.

summary:  [pennant_gdb_log_summary.txt](https://github.com/StanfordLegion/legion/files/10834864/pennant_gdb_log_summary.txt)

full session: [pennant_gdb_log.txt](https://github.com/StanfordLegion/legion/files/10834865/pennant_gdb_log.txt)

I don't see, given these logs, how I could be creating a different set of futures on the second pass through the trace.
